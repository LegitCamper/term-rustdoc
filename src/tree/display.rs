use super::*;
use std::fmt::{self, Debug};

/// skip formatting the field when the value is empty or false
macro_rules! skip_fmt {
    ($base:ident, $self:ident . $($field:ident)+ ) => {$(
        if !$self.$field.is_empty() {
            $base.field(::std::stringify!($field), &$self.$field);
        }
    )+};
    (bool: $base:ident, $self:ident . $($field:ident)+ ) => {$(
        if $self.$field {
            $base.field(::std::stringify!($field), &$self.$field);
        }
    )+};
    (option: $base:ident, $self:ident . $($field:ident)+ ) => {$(
        if $self.$field.is_some() {
            $base.field(::std::stringify!($field), &$self.$field);
        }
    )+};
    (0: $base:ident, $self:ident . $($field:ident)+ ) => {$(
        if $self.$field != 0 {
            $base.field(::std::stringify!($field), &$self.$field);
        }
    )+};
}

impl Debug for DModule {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut base = f.debug_struct("DModule");
        base.field("id", &self.id);
        skip_fmt!(
            base, self . modules structs unions enums
            functions traits constants statics type_alias imports macros
        );
        base.finish()
    }
}

impl Debug for DImpl {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut base = f.debug_struct("DImpl");
        skip_fmt!(base, self . inherent trait_ auto blanket);
        base.finish()
    }
}

impl Debug for DStruct {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut base = f.debug_struct("DStruct");
        base.field("id", &self.id);
        skip_fmt!(bool: base, self.contain_private_fields);
        skip_fmt!(base, self . fields impls);
        base.finish()
    }
}

impl Debug for DUnion {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut base = f.debug_struct("DUnion");
        base.field("id", &self.id);
        skip_fmt!(
            base, self . fields impls
        );
        base.finish()
    }
}

impl Debug for DEnum {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut base = f.debug_struct("DEnum");
        base.field("id", &self.id);
        skip_fmt!(base, self . variants impls);
        base.finish()
    }
}

impl Debug for DTrait {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut base = f.debug_struct("DTrait");
        base.field("id", &self.id);
        skip_fmt!(base, self . types constants functions implementations);
        base.finish()
    }
}

impl Debug for DTypeAlias {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut base = f.debug_struct("DTypeAlias");
        base.field("id", &self.id);
        skip_fmt!(option: base, self.source_path);
        base.finish()
    }
}

impl Debug for TotolCount {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut base = f.debug_struct("TotolCount");
        skip_fmt!(
            0: base, self . modules structs unions enums functions
            traits constants statics type_alias imports
            macros_decl macros_func macros_attr macros_derv
        );
        base.finish()
    }
}
