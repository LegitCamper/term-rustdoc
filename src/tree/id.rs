use crate::util::{xformat, CompactStringExt, XString};
use rustdoc_types::{
    Crate, GenericArg, GenericArgs, Id, Item, ItemEnum, ItemKind, ItemSummary, Path, Type,
};
use std::{borrow::Borrow, cell::RefCell, collections::HashMap};

// NOTE: potentially small improvements on id by using CompactString,
// but this requires the HashMaps to be indexed by other id types:
// * IndexMap<CompactString, Value, FxHash> can improve the hashing speed and
//   get the value from multiple type keys via Equivalent trait
// * so does the tiny improvement really matter?
// #[repr(transparent)]
// pub struct ID {
//     pub id: XString,
// }
pub type ID = String;
pub type IDs = Box<[ID]>;

pub trait IdToID: Sized {
    fn to_ID(&self) -> ID;
    fn into_ID(self) -> ID;
}

impl IdToID for Id {
    fn to_ID(&self) -> ID {
        self.0.clone()
    }

    fn into_ID(self) -> ID {
        self.0
    }
}

impl IdToID for String {
    fn to_ID(&self) -> ID {
        self.clone()
    }

    fn into_ID(self) -> ID {
        self
    }
}

impl IdToID for &str {
    fn to_ID(&self) -> ID {
        (*self).to_owned()
    }

    fn into_ID(self) -> ID {
        self.to_owned()
    }
}

pub trait SliceToIds {
    fn to_ids(&self) -> IDs;
}
impl<T: IdToID> SliceToIds for [T] {
    fn to_ids(&self) -> IDs {
        self.iter().map(|id| id.to_ID()).collect()
    }
}

pub trait IdAsStr {
    fn id_str(&self) -> &str;
}
impl IdAsStr for str {
    fn id_str(&self) -> &str {
        self
    }
}
impl IdAsStr for ID {
    fn id_str(&self) -> &str {
        self
    }
}
impl IdAsStr for Id {
    fn id_str(&self) -> &str {
        &self.0
    }
}
impl<T: IdAsStr> IdAsStr for &T {
    fn id_str(&self) -> &str {
        T::id_str(self)
    }
}

// Crate.index
pub type IndexMap = HashMap<Id, Item>;
// Crate.paths
pub type PathMap = HashMap<Id, ItemSummary>;

pub struct IDMap<'krate> {
    pub index: &'krate IndexMap,
    pub paths: &'krate PathMap,
    id_buffer: RefCell<String>,
}

impl IDMap<'_> {
    pub fn from_crate(krate: &Crate) -> IDMap<'_> {
        IDMap {
            index: &krate.index,
            paths: &krate.paths,
            id_buffer: Default::default(),
        }
    }

    /// Use [`rustdoc_types::Id`] in a buffered way in hot querys.
    pub fn use_id<T>(&self, id: &str, f: impl FnOnce(&Id) -> T) -> T {
        // idbuf always serves as Id used in a query
        let mut buf = self.id_buffer.take();
        buf.clear();
        buf.push_str(id);

        let id = Id(buf);
        let val = f(&id);

        // put the buffer back to be used next time
        self.id_buffer.replace(id.0);

        val
    }
}

/// Get the shortest item name only based on IndexMap.
impl IDMap<'_> {
    pub fn get_item(&self, id: &str) -> Option<&Item> {
        self.use_id(id, |id| self.index.get(id))
    }

    // fn use_item(&self, id: &Id, f: impl FnOnce(&Item) -> XString) -> Option<XString> {
    //     self.get_item(&id.0).map(f)
    // }
    //

    /// If the id doesn't refer to an Item, emit a warn and use the id as the result.
    fn use_item_well(&self, id: &str, f: impl FnOnce(&Item) -> XString) -> XString {
        match self.get_item(id).map(f) {
            Some(s) => s,
            None => {
                warn!("Id({id}) doesn't refer to an Item in IndexMap");
                XString::from(id)
            }
        }
    }

    /// * If the id refers to an Item with a name, use the name;
    /// * if not, try getting the name depending on item type;
    /// * id will be used as the result when name parsing is not yet supported.
    pub fn name<S>(&self, id: &S) -> XString
    where
        S: ?Sized + IdAsStr,
    {
        let id = id.id_str();
        self.use_item_well(id, |item| {
            let name = item.name.as_deref().map(XString::from);
            name.unwrap_or_else(|| item_name(item).unwrap_or_else(|| id.into()))
        })
    }
}

/// Deduce the name from its item type.
fn item_name(item: &Item) -> Option<XString> {
    match &item.inner {
        ItemEnum::Impl(item) => {
            let implementor = item
                .blanket_impl
                .as_ref()
                .and_then(type_name)
                .or_else(|| type_name(&item.for_))
                .unwrap_or_default();
            let trait_ = item.trait_.as_ref().and_then(resolved_path_name)?;
            Some(xformat!("{implementor}: {trait_}"))
        }
        _ => None,
    }
}

const COMMA: XString = XString::new_inline(", ");

fn type_name(ty: &Type) -> Option<XString> {
    match ty {
        Type::ResolvedPath(p) => resolved_path_name(p),
        Type::Generic(t) => Some(t.as_str().into()),
        _ => None,
    }
}

fn resolved_path_name(p: &Path) -> Option<XString> {
    let name = p.name.as_str();
    match p.args.as_deref() {
        Some(GenericArgs::AngleBracketed { args, bindings: _ }) => {
            // FIXME: bindings without args
            if args.is_empty() {
                Some(name.into())
            } else {
                Some(xformat!(
                    "{name}<{}>",
                    args.iter()
                        .filter_map(generic_arg_name)
                        .intersperse(COMMA)
                        .collect::<XString>()
                ))
            }
        }
        Some(GenericArgs::Parenthesized { inputs, output }) => {
            let args = inputs
                .iter()
                .filter_map(type_name)
                .intersperse(COMMA)
                .collect::<XString>();
            let ret = output
                .as_ref()
                .and_then(|t| Some(xformat!(" -> {}", type_name(t)?)))
                .unwrap_or_default();
            Some(xformat!("{name}({args}){ret}"))
        }
        None => Some(name.into()),
    }
}

fn generic_arg_name(arg: &GenericArg) -> Option<XString> {
    match arg {
        GenericArg::Lifetime(life) => Some(life.as_str().into()),
        GenericArg::Type(ty) => type_name(ty),
        GenericArg::Const(_) => None,
        GenericArg::Infer => Some(XString::new_inline("_")),
    }
}

/// Get the external item path only based on PathMap.
impl IDMap<'_> {
    pub fn get_path(&self, id: &str) -> Option<&ItemSummary> {
        self.use_id(id, |id| self.paths.get(id))
    }

    // fn use_path(&self, id: &Id, f: impl FnOnce(&ItemSummary) -> XString) -> Option<XString> {
    //     self.get_path(&id.0).map(f)
    // }

    /// If the id doesn't refer to an ItemSummary, emit a warn and use the id as the result.
    fn use_path_well(&self, id: &str, f: impl FnOnce(&ItemSummary) -> XString) -> XString {
        match self.get_path(id).map(f) {
            Some(s) => s,
            None => {
                warn!("Id({id}) doesn't refer to an Item in IndexMap");
                XString::from(id)
            }
        }
    }

    /// If the id doesn't refer to an ItemSummary with exact given kind,
    /// emit a warn and use the id as the result.
    pub fn path<S, K>(&self, id: &S, kind: K) -> XString
    where
        S: ?Sized + IdAsStr,
        K: Borrow<ItemKind>,
    {
        self.use_path_well(id.id_str(), move |item| {
            let id = id.id_str();
            let kind = kind.borrow();
            if &item.kind == kind {
                item.path.join_compact("::")
            } else {
                warn!(
                    "Id({id}) in PathMap is found as {:?}, but {kind:?} is required",
                    item.kind
                );
                XString::from(id)
            }
        })
    }
}
