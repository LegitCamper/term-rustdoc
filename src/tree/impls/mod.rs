mod debug;

#[macro_use]
pub mod show;
