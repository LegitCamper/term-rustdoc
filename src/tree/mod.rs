#[macro_use]
mod impls;
// The inner macro `icon!` can be used afterwards in submods

#[allow(non_snake_case)]
mod id;
mod nodes;
mod stats;

pub use id::{IDMap, IDs, IdAsStr, IdToID, IndexMap, PathMap, SliceToIds, ID};
pub use impls::show::{DocTree, Show, Tag, TreeLine, TreeLines};
pub use nodes::{
    DConstant, DEnum, DFunction, DImpl, DMacroAttr, DMacroDecl, DMacroDerv, DMacroFunc, DModule,
    DStatic, DStruct, DTrait, DTypeAlias, DUnion,
};
pub use stats::{ImplCount, ImplCounts, ImplKind, ItemCount};
