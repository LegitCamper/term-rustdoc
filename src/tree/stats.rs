use super::{DMacroKind, DModule};
use std::ops::{Add, AddAssign};

#[derive(Default)]
pub struct TotolCount {
    pub modules: u32,
    pub structs: u32,
    pub unions: u32,
    pub enums: u32,
    pub functions: u32,
    pub traits: u32,
    pub constants: u32,
    pub statics: u32,
    pub type_alias: u32,
    pub imports: u32,
    pub macros_decl: u32,
    pub macros_func: u32,
    pub macros_attr: u32,
    pub macros_derv: u32,
}

impl Add for TotolCount {
    type Output = TotolCount;

    fn add(self, rhs: Self) -> Self::Output {
        TotolCount {
            modules: self.modules + rhs.modules,
            structs: self.structs + rhs.structs,
            unions: self.unions + rhs.unions,
            enums: self.enums + rhs.enums,
            functions: self.functions + rhs.functions,
            traits: self.traits + rhs.traits,
            constants: self.constants + rhs.constants,
            statics: self.statics + rhs.statics,
            type_alias: self.type_alias + rhs.type_alias,
            imports: self.imports + rhs.imports,
            macros_decl: self.macros_decl + rhs.macros_decl,
            macros_func: self.macros_func + rhs.macros_func,
            macros_attr: self.macros_attr + rhs.macros_attr,
            macros_derv: self.macros_derv + rhs.macros_derv,
        }
    }
}

impl AddAssign for TotolCount {
    fn add_assign(&mut self, rhs: Self) {
        self.modules += rhs.modules;
        self.structs += rhs.structs;
        self.unions += rhs.unions;
        self.enums += rhs.enums;
        self.functions += rhs.functions;
        self.traits += rhs.traits;
        self.constants += rhs.constants;
        self.statics += rhs.statics;
        self.type_alias += rhs.type_alias;
        self.imports += rhs.imports;
        self.macros_decl += rhs.macros_decl;
        self.macros_func += rhs.macros_func;
        self.macros_attr += rhs.macros_attr;
        self.macros_derv += rhs.macros_derv;
    }
}

impl DModule {
    /// Count the items under current module excluding the current module itself.
    #[rustfmt::skip]
    pub fn current_items_counts(&self) -> TotolCount {
        macro_rules! len {
            ($self:ident . $( $field:ident )+ ) => { $(
                let $field = $self.$field.len().try_into()
                    .expect("the count exceeds the maximum of u32");
            )+ };
        }
        len!(self . modules structs unions enums functions 
             traits constants statics type_alias imports);
        let [mut decl, mut func, mut attr, mut derv]: [u32; 4] = Default::default();
        for m in &self.macros {
            match m.kind {
                DMacroKind::Declarative => decl += 1,
                DMacroKind::ProcFunction => func += 1,
                DMacroKind::ProcAttribute => attr += 1,
                DMacroKind::ProcDerive => derv += 1,
            }
        }
        TotolCount {
            modules, structs, unions, enums, functions, 
            traits, constants, statics, type_alias, imports,
            macros_decl: decl,
            macros_func: func,
            macros_attr: attr,
            macros_derv: derv,
        }
    }

    /// Count all the items excluding the root itself.
    pub fn recursive_items_counts(&self) -> TotolCount {
        self.modules.iter().map(Self::current_items_counts).fold(
            self.current_items_counts(),
            |mut acc, tc| {
                acc += tc;
                acc
            },
        )   
    }
}
