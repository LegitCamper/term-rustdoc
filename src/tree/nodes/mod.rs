mod impls;
use std::ops::Not;

pub use impls::DImpl;

mod structs;
pub use structs::DStruct;

mod unions;
pub use unions::DUnion;

mod enums;
pub use enums::DEnum;

mod traits;
pub use traits::DTrait;

mod type_alias;
pub use type_alias::DTypeAlias;

mod imports;

use super::IDMap;
use crate::tree::{
    impls::show::{DocTree, Show},
    IdToID, IndexMap, ID,
};
use rustdoc_types::{Crate, Id, Item, ItemEnum, ItemKind, MacroKind, Module};

/// Module tree with structural items.
/// All the items only carry ids without actual data.
// NOTE: small improvement by turning all the types of fields
// from Vec to Arr after instantiation.
#[derive(Default)]
pub struct DModule {
    pub id: ID,
    // If true, this module is not part of the public API,
    // but it contains items that are re-exported as public API.
    // is_stripped: bool,
    pub modules: Vec<DModule>,
    pub structs: Vec<DStruct>,
    pub unions: Vec<DUnion>,
    pub enums: Vec<DEnum>,
    pub functions: Vec<DFunction>,
    pub traits: Vec<DTrait>,
    pub constants: Vec<DConstant>,
    pub statics: Vec<DStatic>,
    pub type_alias: Vec<DTypeAlias>,
    pub imports: Vec<()>,
    pub macros_decl: Vec<DMacroDecl>,
    pub macros_func: Vec<DMacroFunc>,
    pub macros_attr: Vec<DMacroAttr>,
    pub macros_derv: Vec<DMacroDerv>,
}

impl DModule {
    pub fn new(doc: &Crate) -> Self {
        // root module/crate name
        let (id, root) = doc
            .index
            .iter()
            .find_map(|(id, item)| {
                if item.crate_id == 0 {
                    if let ItemEnum::Module(Module {
                        is_crate: true,
                        items,
                        ..
                    }) = &item.inner
                    {
                        return Some((id.to_ID(), items.as_slice()));
                    }
                }
                None
            })
            .expect("root module not found");
        Self::new_inner(id, root, &doc.index)
    }

    fn new_inner(id: ID, inner_items: &[Id], index: &IndexMap) -> Self {
        let mut dmod = DModule {
            id,
            ..Default::default()
        };
        dmod.extract_items(inner_items, index);
        dmod
    }

    fn extract_items(&mut self, inner_items: &[Id], index: &IndexMap) {
        for item_id in inner_items {
            match index.get(item_id) {
                Some(item) => self.append(item, index),
                None => warn!("the local item {item_id:?} not found in Crate's index"),
            }
        }
    }

    fn append(&mut self, item: &Item, index: &IndexMap) {
        use ItemEnum::*;
        let id = item.id.to_ID();
        match &item.inner {
            Module(item) => self.modules.push(Self::new_inner(id, &item.items, index)),
            Struct(item) => self.structs.push(DStruct::new(id, item, index)),
            Union(item) => self.unions.push(DUnion::new(id, item, index)),
            Enum(item) => self.enums.push(DEnum::new(id, item, index)),
            Function(_) => self.functions.push(DFunction::new(id)),
            Trait(item) => self.traits.push(DTrait::new(id, item, index)),
            Constant(_) => self.constants.push(DConstant::new(id)),
            Static(_) => self.statics.push(DStatic::new(id)),
            TypeAlias(item) => self.type_alias.push(DTypeAlias::new(id, item)),
            Macro(_) => self.macros_decl.push(DMacroDecl::new(id)),
            ProcMacro(proc) => match proc.kind {
                MacroKind::Bang => self.macros_func.push(DMacroFunc::new(id)),
                MacroKind::Attr => self.macros_attr.push(DMacroAttr::new(id)),
                MacroKind::Derive => self.macros_derv.push(DMacroDerv::new(id)),
            },
            // Primitive(_) => todo!(),
            _ => (),
        }
    }
}

macro_rules! chain {
    ($field:ident, $node:literal, $self:ident, $map:ident) => {
        $self.$field.is_empty().not().then(|| {
            $node
                .show()
                .with_leaves($self.$field.iter().map(|val| val.show_prettier($map)))
        })
    };
}

/// To a recursive tree displayed with ids as nodes.
impl Show for DModule {
    fn show(&self) -> DocTree {
        format!("[mod] {}", self.id).show().with_leaves(
            self.modules.iter().map(DModule::show).chain([
                "Structs"
                    .show()
                    .with_leaves(self.structs.iter().map(DStruct::show)),
                "Unions"
                    .show()
                    .with_leaves(self.unions.iter().map(DUnion::show)),
                "Enums"
                    .show()
                    .with_leaves(self.enums.iter().map(DEnum::show)),
                "Traits"
                    .show()
                    .with_leaves(self.traits.iter().map(DTrait::show)),
                "Functions"
                    .show()
                    .with_leaves(self.functions.iter().map(DFunction::show)),
                "Constants"
                    .show()
                    .with_leaves(self.constants.iter().map(DConstant::show)),
                "Statics"
                    .show()
                    .with_leaves(self.statics.iter().map(DStatic::show)),
                "Macros - Declarative"
                    .show()
                    .with_leaves(self.macros_decl.iter().map(DMacroDecl::show)),
                "Macro - Function"
                    .show()
                    .with_leaves(self.macros_func.iter().map(DMacroFunc::show)),
                "Macro - Attribute"
                    .show()
                    .with_leaves(self.macros_attr.iter().map(DMacroAttr::show)),
                "Macro - Derive"
                    .show()
                    .with_leaves(self.macros_derv.iter().map(DMacroDerv::show)),
            ]),
        )
    }

    fn show_prettier(&self, map: &IDMap) -> DocTree {
        node!("[mod] {}", map.path(&self.id, ItemKind::Module)).with_leaves(
            self.modules
                .iter()
                .map(|m| m.show_prettier(map))
                .chain(chain!(structs, "Structs", self, map))
                .chain(chain!(unions, "Unions", self, map))
                .chain(chain!(enums, "Enums", self, map))
                .chain(chain!(traits, "Traits", self, map))
                .chain(chain!(functions, "Functions", self, map))
                .chain(chain!(constants, "Constants", self, map))
                .chain(chain!(statics, "Statics", self, map))
                .chain(chain!(macros_decl, "Macros - Declarative", self, map))
                .chain(chain!(macros_func, "Macros - Function", self, map))
                .chain(chain!(macros_attr, "Macros - Attribute", self, map))
                .chain(chain!(macros_derv, "Macros - Derive", self, map)),
        )
    }
}

/// generate id wrapper types for simple items
macro_rules! gen_simple_items {
    ($($name:ident => $show:literal -> $kind:ident),+ $(,)?) => {$(
        #[derive(Debug)] pub struct $name { pub id: ID, }
        impl $name { pub fn new(id: ID) -> Self { Self { id } } }
        impl Show for $name {
            fn show(&self) -> DocTree { self.id.show() }
            fn show_prettier(&self, map: &IDMap) -> DocTree {
                node!($show, map.path(&self.id, ItemKind::$kind))
            }
        }
    )+};
}
gen_simple_items! {
    DFunction => "[fn] {}" -> Function ,
    DStatic => "[static] {}" -> Static,
    DConstant => "[constant] {}" -> Constant,
    DMacroDecl => "[macro decl] {}" -> Macro,
    DMacroFunc => "[macro func] {}" -> Macro,
    DMacroAttr => "[macro attr] {}" -> ProcAttribute,
    DMacroDerv => "[macro derv] {}" -> ProcDerive,
}

// TODO:
// *  structs, enums, and enum variants: [non_exhaustive]
//
// [non_exhaustive]: https://doc.rust-lang.org/reference/attributes/type_system.html#the-non_exhaustive-attribute
