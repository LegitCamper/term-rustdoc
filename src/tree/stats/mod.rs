use super::{DMacroKind, DModule};

mod impls;
use impls::acc_sum;

#[derive(Default, Clone)]
pub struct ItemCount {
    pub modules: u32,
    pub structs: u32,
    pub unions: u32,
    pub enums: u32,
    pub functions: u32,
    pub traits: u32,
    pub constants: u32,
    pub statics: u32,
    pub type_alias: u32,
    pub imports: u32,
    pub macros_decl: u32,
    pub macros_func: u32,
    pub macros_attr: u32,
    pub macros_derv: u32,
}

impl DModule {
    /// Count the items under current module excluding the current module itself.
    #[rustfmt::skip]
    pub fn current_items_counts(&self) -> ItemCount {
        macro_rules! len {
            ($self:ident . $( $field:ident )+ ) => { $(
                let $field = $self.$field.len().try_into()
                    .expect("the count exceeds the maximum of u32");
            )+ };
        }
        len!(self . modules structs unions enums functions traits constants statics type_alias imports);
        let [mut decl, mut func, mut attr, mut derv]: [u32; 4] = Default::default();
        for m in &self.macros {
            match m.kind {
                DMacroKind::Declarative => decl += 1,
                DMacroKind::ProcFunction => func += 1,
                DMacroKind::ProcAttribute => attr += 1,
                DMacroKind::ProcDerive => derv += 1,
            }
        }
        ItemCount {
            modules, structs, unions, enums, functions, traits, constants, statics, type_alias, imports,
            macros_decl: decl,
            macros_func: func,
            macros_attr: attr,
            macros_derv: derv,
        }
    }

    /// Count all the items excluding the root itself.
    pub fn recursive_items_counts(&self) -> ItemCount {
        self.modules
            .iter()
            .map(Self::current_items_counts)
            .fold(self.current_items_counts(), acc_sum)
    }
}

/// This type implements `Add` and `AddAssign`, and it means
/// when both operands are the same, the output is the same,
/// if not, the output is `ImplKind::Both`.
///
/// This leads to before you add [`ImplCount`]s, if you care about
/// the same impl kind, you should check it by yourself.
#[derive(Clone, Copy, Debug)]
pub enum ImplKind {
    Inherent,
    Trait,
    Both,
}

#[derive(Clone, Copy)]
pub struct ImplCount {
    pub structs: u32,
    pub enums: u32,
    pub unions: u32,
    pub kind: ImplKind,
    pub total: u32,
}

#[derive(Clone)]
pub struct ImplCounts {
    pub inherent: ImplCount,
    pub trait_: ImplCount,
    pub total: ImplCount,
}

impl ImplCounts {
    pub const EMPTY: Self = ImplCounts {
        inherent: ImplCount {
            structs: 0,
            enums: 0,
            unions: 0,
            kind: ImplKind::Inherent,
            total: 0,
        },
        trait_: ImplCount {
            structs: 0,
            enums: 0,
            unions: 0,
            kind: ImplKind::Trait,
            total: 0,
        },
        total: ImplCount {
            structs: 0,
            enums: 0,
            unions: 0,
            kind: ImplKind::Both,
            total: 0,
        },
    };
}

impl DModule {
    pub fn current_impls_counts(&self) -> ImplCounts {
        let (s_in, s_tr) = {
            let iter = self.structs.iter();
            (
                iter.clone().map(|s| s.impls.inherent.len()).sum::<usize>(),
                iter.map(|s| s.impls.trait_.len()).sum::<usize>(),
            )
        };
        let (e_in, e_tr) = {
            let iter = self.enums.iter();
            (
                iter.clone().map(|s| s.impls.inherent.len()).sum::<usize>(),
                iter.map(|s| s.impls.trait_.len()).sum::<usize>(),
            )
        };
        let (u_in, u_tr) = {
            let iter = self.unions.iter();
            (
                iter.clone().map(|s| s.impls.inherent.len()).sum::<usize>(),
                iter.map(|s| s.impls.trait_.len()).sum::<usize>(),
            )
        };
        let inherent = ImplCount {
            structs: s_in as _,
            enums: e_in as _,
            unions: u_in as _,
            kind: ImplKind::Inherent,
            total: (s_in + e_in + u_in) as _,
        };
        let trait_ = ImplCount {
            structs: s_tr as _,
            enums: e_tr as _,
            unions: u_tr as _,
            kind: ImplKind::Trait,
            total: (s_tr + e_tr + u_tr) as _,
        };
        ImplCounts {
            total: inherent + trait_,
            inherent,
            trait_,
        }
    }

    pub fn recursive_impls_counts(&self) -> ImplCounts {
        self.modules
            .iter()
            .map(Self::current_impls_counts)
            .fold(self.current_impls_counts(), acc_sum)
    }
}
