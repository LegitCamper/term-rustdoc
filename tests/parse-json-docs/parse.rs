use crate::{shot, snap, INTEGRATION};
use similar_asserts::assert_eq;
use term_rustdoc::tree::{DModule, IDMap, Show, TreeLines};

#[test]
fn parse_module() {
    let doc = &INTEGRATION.doc;
    let parsed = DModule::new(doc);
    snap!("DModule", parsed);
    shot!("show-id", parsed.show());

    let tree = parsed.show_prettier(&IDMap::from_crate(doc));
    let display = tree.to_string();
    shot!("show-prettier", display);

    let (flatten_tree, empty) = TreeLines::new(tree);
    let display_new = flatten_tree.display_as_plain_text();
    assert_eq!(expected: display, actual: display_new);
    snap!("flatten-tree", flatten_tree.lines());
    shot!("empty-tree-with-same-depth", empty);

    snap!(parsed.current_items_counts(), @r###"
    ItemCount {
        modules: 1,
        structs: 2,
        functions: 3,
        traits: 1,
        constants: 2,
        macros_decl: 1,
    }
    "###);
    snap!(parsed.recursive_items_counts(), @r###"
    ItemCount {
        modules: 2,
        structs: 2,
        enums: 1,
        functions: 3,
        traits: 1,
        constants: 2,
        macros_decl: 1,
    }
    "###);

    snap!(parsed.current_impls_counts(), @r###"
    ImplCounts {
        total: ImplCount {
            kind: Both,
            total: 3,
            structs: 3,
        },
        inherent: ImplCount {
            kind: Inherent,
            total: 1,
            structs: 1,
        },
        trait: ImplCount {
            kind: Trait,
            total: 2,
            structs: 2,
        },
    }
    "###);
    snap!(parsed.recursive_impls_counts(), @r###"
    ImplCounts {
        total: ImplCount {
            kind: Both,
            total: 5,
            structs: 3,
            enums: 2,
        },
        inherent: ImplCount {
            kind: Both,
            total: 2,
            structs: 1,
            enums: 1,
        },
        trait: ImplCount {
            kind: Trait,
            total: 3,
            structs: 2,
            enums: 1,
        },
    }
    "###);
}
