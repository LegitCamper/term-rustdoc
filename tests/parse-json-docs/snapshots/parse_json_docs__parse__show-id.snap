---
source: tests/parse-json-docs/parse.rs
expression: parsed.show()
---
[mod] 0:0:1800
├── [mod] 0:10:1777
│   ├── [mod] 0:11:1778
│   │   └── Traits
│   │       └── [trait] 0:14:1780
│   │           ├── Associated Types
│   │           ├── Associated Constants
│   │           ├── Associated Functions
│   │           └── Implementors
│   └── Enums
│       └── [enum]
│           ├── Variants
│           │   ├── 0:34:1782
│           │   ├── 0:36:1783
│           │   └── 0:38:106
│           └── Implementations
│               ├── Inherent Impls
│               │   └── 0:15
│               ├── Trait Impls
│               │   └── 0:40
│               ├── Auto Impls
│               │   ├── a:2:2957:254-0:33:1781
│               │   ├── a:2:42204:5740-0:33:1781
│               │   ├── a:2:32574:2059-0:33:1781
│               │   ├── a:2:32492:244-0:33:1781
│               │   └── a:2:42205:5922-0:33:1781
│               └── Blanket Impls
│                   ├── b:2:2745-0:33:1781
│                   ├── b:2:3504-0:33:1781
│                   ├── b:2:2435-0:33:1781
│                   ├── b:2:2750-0:33:1781
│                   ├── b:2:2739-0:33:1781
│                   ├── b:2:2735-0:33:1781
│                   └── b:2:2432-0:33:1781
├── Structs
│   ├── [struct] 0:3:1774
│   │   ├── Fields
│   │   └── Implementations
│   │       ├── Inherent Impls
│   │       ├── Trait Impls
│   │       │   └── 0:6
│   │       ├── Auto Impls
│   │       │   ├── a:2:2957:254-0:3:1774
│   │       │   ├── a:2:42204:5740-0:3:1774
│   │       │   ├── a:2:32574:2059-0:3:1774
│   │       │   ├── a:2:32492:244-0:3:1774
│   │       │   └── a:2:42205:5922-0:3:1774
│   │       └── Blanket Impls
│   │           ├── b:2:2745-0:3:1774
│   │           ├── b:2:3504-0:3:1774
│   │           ├── b:2:2435-0:3:1774
│   │           ├── b:2:2750-0:3:1774
│   │           ├── b:2:2739-0:3:1774
│   │           ├── b:2:2735-0:3:1774
│   │           └── b:2:2432-0:3:1774
│   └── [struct] 0:17:1787
│       ├── Fields
│       │   ├── 0:18:1788
│       │   ├── 0:19:1789
│       │   ├── 0:20:1790
│       │   └── /* private fields */
│       └── Implementations
│           ├── Inherent Impls
│           │   └── 0:22
│           ├── Trait Impls
│           │   └── 0:25
│           ├── Auto Impls
│           │   ├── a:2:2957:254-0:17:1787
│           │   ├── a:2:42204:5740-0:17:1787
│           │   ├── a:2:32574:2059-0:17:1787
│           │   ├── a:2:32492:244-0:17:1787
│           │   └── a:2:42205:5922-0:17:1787
│           └── Blanket Impls
│               ├── b:2:2745-0:17:1787
│               ├── b:2:3504-0:17:1787
│               ├── b:2:2435-0:17:1787
│               ├── b:2:2750-0:17:1787
│               ├── b:2:2739-0:17:1787
│               ├── b:2:2735-0:17:1787
│               └── b:2:2432-0:17:1787
├── Traits
│   └── [trait] 0:5:1775
│       ├── Associated Types
│       ├── Associated Constants
│       ├── Associated Functions
│       └── Implementors
│           └── 0:6
├── Functions
│   ├── 0:27:1793
│   ├── 0:28:1794
│   └── 0:29:1795
├── Constants
│   ├── 0:30:1796
│   └── 0:31:1798
└── Macros - Declarative
    └── 0:32:1799

